server/svix-server/src/error.rs

```diff
@@ -54,53 +78,139 @@ impl error::Error for Error {
     }
 }
 
-impl From<String> for Error {
-    fn from(err: String) -> Error {
-        Error::Generic(err)
+impl IntoResponse for Error {
+    fn into_response(self) -> Response {
+        match self.typ {
+            ErrorType::Http(s) => {
+                tracing::debug!("{:?}", &s);
+                s.into_response()
+            }
+            s => {
+                tracing::error!("type: {:?}, location: {:?}", s, &self.trace);
+                (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({}))).into_response()
+            }
+        }
     }
 }
 
-impl<T: error::Error + 'static> From<bb8::RunError<T>> for Error {
-    fn from(err: bb8::RunError<T>) -> Self {
-        Error::Queue(err.to_string())
-    }
+/// Returns a [&'static str] of the file.rs:<line_number>.
+#[macro_export]
+macro_rules! location {
+    () => {
+        concat!(file!(), ":", line!())
+    };
 }
-impl From<redis::RedisError> for Error {
-    fn from(err: redis::RedisError) -> Self {
-        Error::Queue(err.to_string())
+
+// Kind of boilerplatey, but these macros make it much easier to use the Error type with diagnostic information
+
+#[macro_export]
+macro_rules! err_generic {
+   ($s:expr) => {
+        $crate::error::Error::generic($s, $crate::location!())
+   };
+   ($($arg:tt)*) => {
+        $crate::error::Error::generic(format!($($arg)*), $crate::location!())
+   }
+}
+
+#[macro_export]
+macro_rules! err_database {
+    ($s:expr) => {
+        $crate::error::Error::database($s, $crate::location!())
+    };
+}
+
+#[macro_export]
+macro_rules! err_queue {
+    ($s:expr) => {
+        $crate::error::Error::queue($s, $crate::location!())
+    };
+}
+
+#[macro_export]
+macro_rules! err_validation {
+    ($s:expr) => {
+        $crate::error::Error::validation($s, $crate::location!())
+    };
+}
+
+pub trait Traceable<T> {
+    fn trace(self, location: &'static str) -> Result<T>;
+}
+
+/// Adds [location!] data to the given result, returning a [crate::error::Result<T>].
+#[macro_export]
+macro_rules! ctx {
+    ($res:expr) => {
+        $crate::error::Traceable::trace($res, $crate::location!())
+    };
+}
+
+impl<T> Traceable<T> for Result<T> {
+    fn trace(self, location: &'static str) -> Result<T> {
+        self.map_err(|mut e| {
+            e.trace.push(location);
+            e
+        })
     }
 }
 
-impl From<DbErr> for Error {
-    fn from(err: DbErr) -> Error {
-        if let DbErr::Query(ref err_str) = err {
-            // Surely there's a better way
-            if err_str.contains("duplicate key value violates unique constraint") {
-                Error::Http(HttpError::conflict(None, None))
+impl<T> Traceable<T> for std::result::Result<T, DbErr> {
+    fn trace(self, location: &'static str) -> Result<T> {
+        self.map_err(|err| {
+            let typ = if let DbErr::Query(ref err_str) = err {
+                // Surely there's a better way
```

Comment:
Without changes to SeaORM, I don't think there is one. I'd probably remove the comment.